### ISSUE TITLE
About the Dataloader Shuffle / Sampler

### ISSUE THREAD

**Issue by zw615** [young-geng/m3ae_public#5]
Hi, I have played with the code for some time. And I find that shuffle argument when creating pytorch dataloader is set to False (e.g. in [pretrain](https://github.com/young-geng/m3ae_public/blob/e3557b4c3370810bc480d4a1c346d84788921969/m3ae/m3ae_main.py#L48) or [finetune](https://github.com/young-geng/m3ae_public/blob/e3557b4c3370810bc480d4a1c346d84788921969/m3ae/finetune_main.py#L48) code). Correct me if I am wrong, but shouldn't this lead to the same data reading order for every epoch? Moreover, considering the fact that the hdf5 file is created based on the order of ImageNet classes as in this [issue](https://github.com/young-geng/m3ae_public/issues/4), when `shuffle=False`, mostly a batch only contain data from a certain class, leading to extremely biased gradient estimation.

More surprisingly, I tried to set `dataloader_shuffle=True`, but only found that the training time increased by about an order of magnitude. I do not have an exact number, as the tqdm estimatation are always changing, probably due to the Asynchronous dispatch nature of JAX. 

To the best of my knowledge, when doing distributed training, it is common practice to use `DistributedSampler` in pytorch dataloader, rather than `shuffle`, as in this official example [here](https://github.com/pytorch/examples/blob/40289773aa4916fad0d50967917b3ae8aa534fd6/imagenet/main.py#L255). I have thought that maybe not using DistributedSampler results in every process reading the whole dataset, instead of a subset. So I have tried switching to DistributedSampler as in the official example, only to find still slow training speed.

I wonder how to fix this issue? Am I missing something here? Thanks a lot!

**Comment by young-geng** [comment]
You are right that it reads the dataset sequentially. For CC12M this is not a problem because CC12M does not group the images into categories. For finetuning on ImageNet this indeed poses a problem where all the images from a single batch belong to the same category. To mitigate this, you'll need to shuffle the ImageNet dataset a head of time. Here's a script that can do that for you:

```python
import os
import tarfile
import json
import random
import numpy as np
from tqdm import tqdm, trange
import h5py



if __name__ == '__main__':
    base_dir = './data/download/download_and_prepare_imagenet_dataset/'

    with open('./data/download/download_and_prepare_imagenet_dataset/imagenet_class_index.json') as fin:
        label_data = json.load(fin)
    labels = [(i, n) for i, (n, _) in label_data.items()]

    # labels = sorted(os.listdir(os.path.join(base_dir, 'train')))

    h5_bytes = h5py.special_dtype(vlen=np.dtype('uint8'))
    h5_file = h5py.File('./imagenet.hdf5','x')

    h5_file.create_dataset('train_jpg', (1400000, ), dtype=h5_bytes, chunks=True)
    h5_file.create_dataset('train_labels', (1400000, ), dtype=np.int32, chunks=True)

    h5_file.create_dataset('val_jpg', (140000, ), dtype=h5_bytes, chunks=True)
    h5_file.create_dataset('val_labels', (140000, ), dtype=np.int32, chunks=True)

    total_train = 0
    total_val = 0

    for label_id, label in tqdm(labels, ncols=0):
        for image in os.listdir(os.path.join(base_dir, 'train', label)):
            if not (image.lower().endswith('.jpeg') or image.lower().endswith('.jpg')):
                continue

            with open(os.path.join(base_dir, 'train', label, image), 'rb') as fin:
                h5_file['train_jpg'][total_train] = np.frombuffer(fin.read(), dtype='uint8')
            h5_file['train_labels'][total_train] = label_id
            total_train += 1

        for image in os.listdir(os.path.join(base_dir, 'val', label)):
            if not (image.lower().endswith('.jpeg') or image.lower().endswith('.jpg')):
                continue

            with open(os.path.join(base_dir, 'val', label, image), 'rb') as fin:
                h5_file['val_jpg'][total_val] = np.frombuffer(fin.read(), dtype='uint8')
            h5_file['val_labels'][total_val] = label_id
            total_val += 1


    h5_file['train_jpg'].resize((total_train,))
    h5_file['train_labels'].resize((total_train,))
    h5_file['val_jpg'].resize((total_val,))
    h5_file['val_labels'].resize((total_val,))


    h5_file_shuffled = h5py.File('./imagenet_shuffled.hdf5','x')

    h5_file_shuffled.create_dataset('train_jpg', (total_train, ), dtype=h5_bytes, chunks=True)
    h5_file_shuffled.create_dataset('train_labels', (total_train, ), dtype=np.int32, chunks=True)

    h5_file_shuffled.create_dataset('val_jpg', (total_val, ), dtype=h5_bytes, chunks=True)
    h5_file_shuffled.create_dataset('val_labels', (total_val, ), dtype=np.int32, chunks=True)


    train_indices = list(range(total_train))
    random.shuffle(train_indices)

    val_indices = list(range(total_val))
    random.shuffle(val_indices)

    for i, k in enumerate(tqdm(train_indices, ncols=0)):
        h5_file_shuffled['train_jpg'][i] = h5_file['train_jpg'][k]
        h5_file_shuffled['train_labels'][i] = h5_file['train_labels'][k]

    for i, k in enumerate(tqdm(val_indices, ncols=0)):
        h5_file_shuffled['val_jpg'][i] = h5_file['val_jpg'][k]
        h5_file_shuffled['val_labels'][i] = h5_file['val_labels'][k]

    h5_file.close()
    h5_file_shuffled.close()

```

**Comment by zw615** [comment]
I guess for larger dataset cc12m (also, which is  not grouped by category), non-shuffling is not a big issue. But for imagenet, even this pre-shuffling when creating the dataset does not really solve the problem, because it only returns the same order for every epoch. In practice, PyTorch DistributedSampler uses the `set_epoch` method to create different shuffling order for different epochs. I will try to use webdataset or tfrecord instead. But thanks a lot anyway!

**Comment by young-geng** [comment]
I guess in practice, we've never noticed this to give us any trouble for ImageNet. We are able to reproduce the MAE paper's results for pre-training then linear probing and fine-tuning on ImageNet.
